src/pmfs/vos_target_fs.c

```diff
@@ -0,0 +1,1180 @@
+
+	rc = pmfs_scan_cont(&ctx, &uoi, opc);
+	if (rc != 0) {
+		D_ERROR("Object get dkeys info failed: " DF_RC "\n",
+			DP_RC(rc));
+		return rc;
+	}
+
+	*nr = uoi->nr;
+	*len = uoi->len;
+	kds = uoi->kds;
+	buf = uoi->buf;
+
+	return 0;
+}
+
+
+struct vos_fs_ult_arg {
+	void *args;
+	enum task_op opc;
+	double *duration;
+	int status;
+};
+
+static int
+_vos_parse_commands(enum task_op opc, void *union_args)
+{
+	int rc = 0;
+
+	if (opc == VOS_OBJ_UPDATE) {
+		struct vos_client_obj_rw_args *args = union_args;
+
+		rc = vos_obj_update(args->coh, args->oid,
+				    args->epoch, args->pm_ver,
+				    args->flags, args->dkey,
+				    args->iod_nr, args->iods, NULL,
+				    args->sgls);
+
+	} else if (opc == VOS_OBJ_FETCH) {
+		struct vos_client_obj_rw_args *args = union_args;
+
+		rc = vos_obj_fetch(args->coh, args->oid,
+				   args->epoch, args->flags,
+				   args->dkey, args->iod_nr,
+				   args->iods, args->sgls);
+	} else if (opc == VOS_OBJ_PUNCH) {
+		struct vos_client_obj_rw_args *args = union_args;
+
+		rc = vos_obj_punch(args->coh, args->oid,
+				   args->epoch, args->pm_ver,
+				   args->flags, args->dkey,
+				   args->akey_nr, args->akeys,
+				   NULL);
+
+	} else if (opc == VOS_OBJ_GET_NUM_DKEYS) {
+		struct vos_client_obj_list_args *args = union_args;
+
+		rc = vos_obj_get_info_dkeys(args->coh, args->oid,
+					    opc, args->nr, args->len,
+					    args->kds, args->buf);
+	} else if (opc == VOS_OBJ_LIST_DKEYS) {
+		struct vos_client_obj_list_args *args = union_args;
+
+		rc = vos_obj_list_info_dkeys(args->coh, args->oid,
+					     opc, args->nr, args->len,
+					     args->kds, args->buf);
+	} else {
+		  D_ERROR("Command is not recongnized.");
+		  rc = -1;
+	}
+
+	return rc;
+}
+
+static int
+_vos_fs_execute_command(void *_args, enum task_op opc, double *duration)
+{
+	double then, now;
+	int rc = 0;
+
+	then = dts_time_now();
+	rc = _vos_parse_commands(opc, _args);
+
+	now = dts_time_now();
+	*duration = now - then;
+#if DEBUG
+	D_PRINT("execute command %s end success\r\n", cmd_string[opc]);
+	D_PRINT("Dts last time = %-10.6fs\r\n", now - then);
+#endif
+	return rc;
+}
+
+static void
+vos_fs_execute_command_ult(void *arg)
+{
+	struct vos_fs_ult_arg *vfua = arg;
+
+	D_INFO("execute command %s\r\n", cmd_string[vfua->opc]);
+	vfua->status =
+		_vos_fs_execute_command(vfua->args, vfua->opc,
+					vfua->duration);
+}
+
+int
+pmfs_thread_create(void *fs_cb, void *arg)
+{
+	int rc;
+
+	ABT_thread thread;
+	ABT_xstream xstream;
+
+	ABT_xstream_create(ABT_SCHED_NULL, &xstream);
+	ABT_xstream_get_main_pools(xstream, 1, &pool);
+
+	rc = ABT_thread_create(pool, fs_cb, arg, ABT_THREAD_ATTR_NULL, &thread);
+	if (rc != ABT_SUCCESS) {
+		D_ERROR("ABT_thread_create failed: " DF_RC "\n",
+			DP_RC(rc));
+		return rc;
+	}
+
+	rc = ABT_thread_join(thread);
+	if (rc != ABT_SUCCESS) {
+		D_ERROR("ABT_thread_join failed:" DF_RC "\n",
+			DP_RC(rc));
+		return rc;
+	}
+
+	ABT_thread_free(&thread);
+	ABT_xstream_free(&xstream);
```

Comment:
Creating a xstream (pthread) for each VOS operation and execute the operation one by one? I bet the performance will be really bad.

I think the threading model should be same to the DAOS engine threading model: Single processing xstream dequeue the item from lockless ring and create dedicated ULT to execute the VOS operation, also, the internal GC & aggregation ULTs are necessary.